portal-bridge/src/gossip.rs

```diff
@@ -95,15 +100,52 @@ pub async fn gossip_history_content(
     content_key: HistoryContentKey,
     content_value: HistoryContentValue,
     block_stats: Arc<Mutex<HistoryBlockStats>>,
+    // Indicates what gossip mode is being used
+    // if None, then the gossip mode is 'gossip'
+    // if Some, then the gossip mode is 'offer'
+    census_tx: Option<tokio::sync::mpsc::UnboundedSender<EnrsRequest>>,
 ) -> anyhow::Result<()> {
-    let result = tokio::spawn(
-        history_trace_gossip(portal_client, content_key.clone(), content_value).in_current_span(),
-    )
-    .await?;
-    if let Ok(mut data) = block_stats.lock() {
-        data.update(content_key, result.into());
+    if let Some(census_tx) = census_tx {
+        let (resp_tx, resp_rx) = futures::channel::oneshot::channel();
+        let enrs_request = EnrsRequest {
+            content_key: ContentKey::History(content_key.clone()),
+            resp_tx,
+        };
+        census_tx.send(enrs_request)?;
+        let enrs = resp_rx.await?;
+        let total_enrs = enrs.len();
+        let mut success = 0;
+        for enr in enrs {
```

Comment:
One thing to consider... When running the bridge in "latest" mode, performance (measured by how many successful transfers there were ... indicating that this mode is actually better / faster at distributing content than the 'gossip' mode) really only was acceptable when I filtered out all Shisui nodes from `enrs`. It seems like we still have some utp tx issues with Shisui. So, this loop would be blocked whenever it encountered a problematic transfer. And by the time the following requests were actually made, the network had been saturated with content from the devops (aka "gossip") bridges.

One solution is to spawn tasks for each `trace_offer`, but that makes tracking the number of concurrent rpc requests (for the gossip limit) problematic. 

Imo, we **should** probably always have bridges running in both modes (eg. 'gossip' & 'offer'). And given that this is really only an issue in "latest" mode, I don't think it's worth hardcoding the Shisui exception.

Reply:
> One solution is to spawn tasks for each trace_offer, but that makes tracking the number of concurrent rpc requests (via the gossip_semaphore) problematic.

I am not sure I see where this sentiment is coming from.

> tracking the number of concurrent rpc requests (via the gossip_semaphore) problematic.

I don't think this would be problem-matic, I don't think a gossip semaphore limit should be used on serving latest data.

I think we should limit how many Enr's we gossip to though as currently their is no limit which is a bit excess work

Reply:
> Imo, we should probably always have bridges running in both modes (eg. 'gossip' & 'offer'). And given that this is really only an issue in "latest" mode, I don't think it's worth hardcoding the Shisui exception

If this is the case I think `offer` mode isn't implemented correctly.

> So, this loop would be blocked whenever it encountered a problematic transfer

There shouldn't be blocking code at that level at least.

> Imo, we should probably always have bridges running in both modes (eg. 'gossip' & 'offer')

A good `offer` implementation I think it only makes sense to bridge in `offer` mode.

All offer mode is, is an optimized version of `gossip` where the bridge has better information for knowing who to gossip to. With better data the bridge can send offer requests in a less wasteful way.

I pointed it out earlier, but I don't think it makes sense to offer every node we are aware of which will accept it. The bridge's job shouldn't be seeding every node on the network, won't the nodes just repeat what they got to their interested neighborhood

Reply:
`Imo, we should probably always have bridges running in both modes (eg. 'gossip' & 'offer')`

I disagree with this. If we have a good `offer mode` implementation a naive `gossip` implementation can't compete on performance or reliability especially as the network grows where bigger and bigger issues will be displayed by `gossip mode`

`gossip mode` will use less ram so it could be helpful if someone wants to contribute with limited hardware, but other then that I think the points against `offer mode` can easily be fixed and then it will just outperform `gossip` mode

Reply:
> I disagree with this. If we have a good `offer mode` implementation a naive `gossip` implementation can't compete on performance or reliability especially as the network grows where bigger and bigger issues will be displayed by `gossip mode`

The issue I have with this is that this then tends to go in the direction of relying only on a "few" mega bridges that can/will do all the work. If these then get attacked, the network stalls in its data injections.

I'm also not so sure if it is the good way forward if we eventually want to end up with Ethereum nodes that do this work and where the general network takes the load of gossiping things around.

Reply:
> > I disagree with this. If we have a good `offer mode` implementation a naive `gossip` implementation can't compete on performance or reliability especially as the network grows where bigger and bigger issues will be displayed by `gossip mode`
> 
> 
> 
> The issue I have with this is that this then tends to go in the direction of relying only on a "few" mega bridges that can/will do all the work. If these then get attacked, the network stalls in its data injections.
> 
> 

You would easily be able to limit the routing table size if that is issue, for lower end machines which wanted to run a bridge. But still allowing the bridge to scale for hardware which can do more is very important.

So I don't think this design decision limits who can bridge, it just allows people with better hardware to utilize it to bridge more efficiently which below I will go over why that is important 

> 
> I'm also not so sure if it is the good way forward if we eventually want to end up with Ethereum nodes that do this work and where the general network takes the load of gossiping things around.
> 
> 

These outlines are meant to meet the needs of the State Network, which has drastically higher requirements than gossiping for the History or Beacon Network. By drastically I mean order of magnitudes.

In the state network full proofs are required to be gossiped for every changed trie node every block. As outlined during the Prague Summit.

So the general network isn't really able to take on massive burden, other than making sure its friends have the data it was gossiped. The bridge takes on a majority of the work.

On top of all this we need to account for executing the block and generating the MPT which is costly then gossip both the all the account trie and storage trie intermediate nodes in under 12 seconds ideally less. Which is no easy feat. It is unrealistic 1 bridge will be able to do this so we will likely have to split how much data is gossiped over many bridges which is expensive.


Then the other issue is for backfilling non-latest data which with the current bridge design I don't want to imagine how long that would take.

Then the third issue is how do we reliably fill holes in the network this is very hard to do, because finding the hole in the first place is a probabilistic anomaly.


So running the state network and maintaining it is very expensive and time consuming, having an efficient bridge is hence important even then it will be super expensive and time consuming. I don't think this bridge design limits lower end nodes from contributing to gossip data, it doesn't limit us in anyway like that.

It realistically enables us to hopefully sustainably run the State Network, which I don't think its requirements can be understated.

For other networks the requirements I outlined are most definitely overkill, but I am not that concerned with other networks, because if we can sustainably make the state network work, those networks are in relation a drop in the bucket.

Reply:
To be clear, my remark was about only relying on the "optimized" bridge architecture. Not about actually having those around for now.

Reply:
> To be clear, my remark was about *only* relying on the "optimized" bridge architecture. Not about actually having those around for now.

Maybe I should have said for State it only makes sense to rely on X to be more clear, sorry for the confusion